DM-38190: Write up discussion about failure modes in Prompt Processing (#1)
lsst-dm/dmtn-260: index.rst

   **This technote is a work-in-progress.**

.. _abstract:

Abstract
========

The Prompt Processing system will be responsible for processing roughly a thousand visits per night, and distributing the results in near real time, for at least ten years of Rubin Observatory operations.
As such, it must be highly robust to algorithmic, network, and infrastructure failures, ranging from momentary glitches to extended downtimes.
`DMTN-219`_ introduced the initial design for the Prompt Processing framework; this document expands on the design to address expected failure modes and recovery strategies for each.

.. _DMTN-219: https://dmtn-219.lsst.io/

.. _pp-overview:

Prompt Processing System Overview
=================================

> The Prompt Processing system will receive images from both LATISS and LSSTCam, and run one of several pipelines on them, depending on the purpose of the observations.

Reviewer: LSSTComCam now, too.

> In the baseline case (running the full Alert Production pipeline on images for the Large Survey of Space and Time), the system will create multiple outputs: the transient alert stream, the alert production database (APDB), datasets in a central repository, and Sasquatch metrics for pipeline analytics and debugging.

Reviewer: ```suggestion
In the baseline case (running the full Alert Production pipeline on images for the Legacy Survey of Space and Time), the system will create multiple outputs: the transient alert stream, the alert production database (APDB), datasets in a central repository including the prompt data products described in the DPDD, and Sasquatch metrics for pipeline analytics and debugging.
```

As of May 2023, the Prompt Processing system starts processing when it receives a ``next_visit`` event from the summit before observations are taken.
The event is received by a fan-out service, which forwards a derived event for each detector to a pool of workers.
Each fanned-out event represents the exposure(s) for a given visit with a given detector.
Because the workers handle their assigned exposures independently and in parallel, this framework is efficient and flexible with respect to unexpected computational demands.
However, it's impossible to make assumptions about the order of events between workers, and any given worker may be created or destroyed at any time.

Each of the several hundred workers maintains an internal ("local") Butler repository for its processing; this keeps the central repository from becoming a performance bottleneck for pipeline execution.
On receiving its visit-detector assignment, each worker downloads any calibrations, templates, etc. it will need but does not yet have from the central repo.
The worker then waits for the raw exposures to arrive at USDF (or confirms they already arrived) and downloads and ingests them.

> When all exposures have arrived, the worker runs the pipeline, updating the APDB and sending alerts as appropriate; if not all exposures arrive, it attempts to process what it has.

Reviewer: How does it detect when to give up?  Must be a timeout, right?

Author: In practice, probably. But is it a relevant detail in what is already a fairly verbose section?

Reviewer: I think it's somewhat important to call it out, as pipeline-level timeouts may interact with worker-level timeouts.

> Once pipeline processing is complete, the worker uploads its outputs to the central repository.

Reviewer: Won't alerts go out directly from the pipeline (not mediated via the central repository)?

Author: Yes. Which part of the wording is misleading?

Reviewer: I was considering alerts to be pipeline outputs, and so it was confusing to think of them as being uploaded to the central repository after pipeline processing, especially in the context of the previous sentence where they're sent "during" the pipeline.

Author: "output datasets"? "Butler products"?

Reviewer: "output datasets" would be fine.

The worker will likely also compute metrics at this stage, but the details (in particular, whether they are computed before or after the central repository sync) have not yet been decided.

For performance reasons, each worker places pipeline outputs from all visits into a small number of Butler runs, which are persisted when each worker uploads to the central repository.
The run name is a deterministic function of the Science Pipelines configuration, the instrument and choice of pipeline, and the processing date.
The use of a separate run for each pipeline is required by the Gen 3 provenance system (in particular, the storage of pipeline configuration information and "init-output" datasets), but makes it difficult for the Prompt Processing system to adjust its choice of pipeline in response to missing inputs or to processing failures.

.. _general-priorities:

General Priorities
==================

Of the outputs produced by the Alert Production pipeline, the most important to have correct and uncorrupted is the APDB, which serves as the source of truth both for later processing and for the externally available Prompt Products database.

> The alert stream is a close second, since it is the primary user interface for transient follow-up and cannot be "fixed" after the fact.

Reviewer: We can reissue or retract alerts, as described below, but the original erroneous alert could have triggered downstream actions.

> The central repository and metrics are less important, because they are (almost) exclusively for internal use.

Reviewer: PVIs, diffims, and catalog Parquet files from AP go to science users (after embargo periods).

Author: Sorry, can you elaborate on the Parquet catalogs? I checked the DPDD (section 3.3), and it only mentions catalog information inside the PPDB itself.

As far as possible, the Prompt Processing pipeline should be idempotent -- in other words, multiple tries of the same pipeline on the same dataset should yield the same results, no matter what the state of the system.
This is possible when rebroadcasting the alert stream -- in the worst case, duplicate alerts exist but have the same position and time -- and when running the pipeline from ISR through image differencing.
However, for reasons explored in :ref:`association`, idempotency cannot be safely achieved with source association.

.. _retries:

Retrying Processing
===================

The execution framework can automatically recover from failures, in the sense of restoring full processing capacity without compromising the framework's state.
However, recovering the data processing itself is the responsibility of the code running on each worker.

Visit processing can fail at any point, including within and between tasks, and while performing external I/O.

> Even within a single worker, the order of events is not fully deterministic (e.g., two snaps could be ingested and processed up to snap combination in parallel).

Reviewer: Could mention that we don't currently enable parallel processing but we might to improve latency.

However, the possible failures can be divided into a handful of cases:

> - Any failure before the end of APDB writing (which is atomic as of May 2023) has no consequences beyond that worker's local repository.

Reviewer: It's good (necessary even) to point out the atomicity of APDB updates, but it might be good to also insert a forward reference here to the difficulties of making them idempotent.

  The repository may have ingested raws, or partial pipeline outputs.
- A failure after APDB writing but before the end of ``DiaPipelineTask``, which contains it, is recoverable only if the task can account for the fact that a previous session already wrote to the APDB.
- A failure after APDB writing but before or during alert distribution cannot be aborted without leaving the APDB inconsistent with the alert stream.
  However, retrying is possible, and easy if duplicate alerts are acceptable (see :ref:`general-priorities`).

> - A failure after alert distribution but but before or during central repo sync may leave the latter missing calexps and difference images, which are supposed to be exported to the Rubin Science Platform.

Reviewer: ```suggestion
- A failure after alert distribution but before or during central repo sync may leave the latter missing PVIs and difference images, which are supposed to be exported to the Rubin Science Platform.
```

- A failure after central repo sync may interfere with internal metrics and monitoring, but has no consequences for science users.

> In general, the Rubin Observatory project has followed a policy of processing data at most once, not at least once.

Reviewer: I'm not sure this is always the case, but for AP in particular, dropping a few visits can still allow the 98% requirement to be met, so we indeed have erred on the side of "at most once".

We propose to do the same thing in Prompt Processing, by attempting to retry processing only in the case of specific failure modes that we know to be recoverable.
As a first approximation, this means that retries are allowed before writing to the APDB, but not afterward.
Any failures that are not retried automatically can still be handled in next-day processing.

.. _association:

APDB and Source Association
===========================

The persistent nature of the APDB makes it difficult to retry processing runs that modify it.

> For example, if DiaSource and DiaObject IDs are deterministic functions of only their visit, then pipeline code might handle retries by testing for these IDs in the APDB, and ignoring or overwriting them.

Reviewer: Shouldn't we already know how DiaSource and DiaObject IDs are generated?

Author: Is the generation algorithm set in stone? I'm trying to avoid justifying any decisions with details that may change later.

Reviewer: I guess things might be non-deterministic if we, for example, parallelized detection over sections of the detector image.  So this is fine.

For example, preventing a retry or a delayed processing run from using any APDB entries added after the "correct" time can lead to two visits creating DiaObjects at the same position because each is required to ignore the other.
More complex strategies using validity ranges or other tools can avoid such paradoxes, but may lead to more subtle bugs.

The simplest way to keep a consistent association order when recovering from processing failures is to allow all runs to use the state of the APDB at the final processing time.
If we (and science users) think of out-of-order visits as precoveries, then there shouldn't be any confusion over the processing order not being strictly chronological.

.. _consistency:

Inconsistent Output
===================

As noted in :ref:`general-priorities`, the most important pipeline outputs are, in order, the APDB, the alert stream, and the central repository and metrics.
As of May 2023, this is also the order in which the pipeline produces outputs.
If pipeline processing fails in its late stages, these outputs may be inconsistent with each other; for example, the APDB may contain DiaSources for which alerts were never sent.

> The alert stream can be easily restored from the APDB, if we have a record of which alerts were sent or can tolerate duplicates.

Reviewer: The Alert Database maintains a record of which alerts were sent.

Such a strategy is safe so long as no code assumes that the existence of one dataset implies the existence of another.
Second, the central repository sync itself may fail, leaving an undefined subset of datasets transferred.
Again, the immediate risk is that something might assume the repository contains a self-consistent set of outputs; in the longer term, the datasets can be regenerated through daytime processing.

.. _bad-data:

Corrupted Pipeline Outputs
==========================

It's possible that some processing errors will allow the pipeline to run to completion while producing large numbers of invalid sources.
Such sources will clutter the alert stream with false positives, and may confuse source association on later visits.

The Prompt Processing framework does not itself have any way to detect nonsense output.
However, the Alert Production team is incorporating "circuit breaker" checks into the pipeline; see `DM-35645`_ and its follow-up issues.
These checks will escalate suspicious outputs into pipeline failures, which can be handled as described above.

> As of May 2023, the proposed checks focus on poor-quality raw inputs; there are no checks specifically guarding DiaSouce detection or the APDB.

Reviewer: ```suggestion
As of May 2023, the proposed checks focus on poor-quality raw inputs; there are no checks specifically guarding DiaSource detection or the APDB.
```

Author: @ebellm, is this statement still true? [DM-35645](https://jira.lsstcorp.org/browse/DM-35645) was resolved in November 2022, but I don't see a "final" list of checks either there or on [DM-37142](https://jira.lsstcorp.org/browse/DM-37142).

Reviewer 2: It still remains to do, but I think the statement that we can have the pipeline raise on bad data is fine.

Author: My question was about "the proposed checks focus on poor-quality raw inputs; there are no checks specifically guarding DiaSource detection or the APDB" -- is that still true, or an overgeneralization?

Reviewer 2: Yes, that's true as of today.

.. _DMTN-259: https://dmtn-259.lsst.io/

.. _timeout:

Pipeline Timeouts
=================

Another risk is that, under some circumstances, the pipeline code may fail to terminate.
This risk is mitigated by the scalable processing framework; taking a single worker out of action will not interfere with the processing of other images or with overall system performance.

The obvious way to handle stuck pipelines is to impose a timeout on either the job or the worker.
The former carries the risk that the worker will be left in an inconsistent state, corrupting future processing, while the latter loses all intermediate data and adds the overhead of starting a new worker and preparing its local repository.
In either case, the possibility of a timeout needs to be accounted for when solving the problems discussed in :ref:`association` and :ref:`consistency`, since a timeout can interrupt processing at *any* time, including during I/O.

It might be possible to impose a timeout at the task level rather than the worker level, to better distinguish between a slow job and a stuck one.

> However, no current framework allows tasks to be timed in real time; for example, timing metrics are only available after a task has completed.

Reviewer: There are reasons for DRP to impose timeouts as well; it's possible that these might be pushed to the task level rather than the workflow job level in a future iteration of the middleware.
